featurebyte/feast/service/registry.py

```diff
@@ -52,11 +55,93 @@ def __init__(  # pylint: disable=too-many-arguments
         self.catalog_service = catalog_service
         self.entity_lookup_feature_table_service = entity_lookup_feature_table_service
 
+    async def _create_project_name(
+        self, catalog_id: ObjectId, hex_digit_num: int = 7, max_try: int = 100
+    ) -> str:
+        # generate 7 hex digits
+        project_name = str(catalog_id)[-hex_digit_num:]
+        document_dict = await self.persistent.find_one(
+            collection_name=self.collection_name,
+            query_filter={"name": project_name},
+            projection={"_id": 1},
+            user_id=self.user.id,
+        )
+        if not document_dict:
+            return project_name
+
+        count = 0
+        while True:
+            project_name = f"{random.randrange(16**hex_digit_num):0{hex_digit_num}x}"
+            document_dict = await self.persistent.find_one(
+                collection_name=self.collection_name,
+                query_filter={"name": project_name},
+                projection={"_id": 1},
+                user_id=self.user.id,
+            )
+            if not document_dict:
+                return project_name
+
+            count += 1
+            if count > max_try:
+                raise RuntimeError("Unable to generate unique project name")
+
+    async def _create_offline_table_name_prefix(self, feature_store_id: ObjectId) -> str:
+        res, _ = await self.persistent.aggregate_find(
+            collection_name=self.collection_name,
+            pipeline=[
+                {"$match": {"feature_store_id": feature_store_id}},
+                {"$group": {"_id": None, "unique_names": {"$addToSet": "$name"}}},
+            ],
+        )
+        results = list(res)
+        found_names = set(results[0]["unique_names"]) if results else set()
+        name_count = len(found_names)
+        return f"cat{name_count + 1}"
+
+    async def get_or_create_feast_registry(
+        self,
+        catalog_id: ObjectId,
+        feature_store_id: Optional[ObjectId],
+    ) -> FeastRegistryModel:
+        """
+        Get or create project name
+
+        Parameters
+        ----------
+        catalog_id: ObjectId
+            Catalog id
+        feature_store_id: Optional[ObjectId]
+            Feature store id
+
+        Returns
+        -------
+        FeastRegistryModel
+        """
+        with self.allow_use_raw_query_filter():
+            query_filter = {"catalog_id": catalog_id}
+            if feature_store_id:
+                query_filter["feature_store_id"] = feature_store_id
+            else:
+                catalog = await self.catalog_service.get_document(document_id=catalog_id)
+                query_filter["feature_store_id"] = catalog.default_feature_store_ids[0]
+
+            query_result = await self.list_documents_as_dict(
+                query_filter=query_filter, use_raw_query_filter=True, page_size=1
+            )
```

Comment:
If the `query_filter` always includes `catalog_id`, is it still necessary to set `use_raw_query_filter`?

Reply:
Good point, will revise this part.